server/src/main/java/org/elasticsearch/index/analysis/ReloadableCustomAnalyzer.java

```diff
@@ -0,0 +1,164 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.index.analysis;
+
+import org.apache.lucene.analysis.Analyzer;
+import org.apache.lucene.analysis.TokenStream;
+import org.apache.lucene.analysis.Tokenizer;
+import org.apache.lucene.util.CloseableThreadLocal;
+import org.elasticsearch.common.settings.Settings;
+
+import java.io.Reader;
+import java.util.Map;
+
+public final class ReloadableCustomAnalyzer extends Analyzer implements AnalyzerComponentsProvider {
+
+    private volatile AnalyzerComponents components;
+
+    private CloseableThreadLocal<AnalyzerComponents> storedComponents = new CloseableThreadLocal<>();
+
+    private final int positionIncrementGap;
+
+    private final int offsetGap;
+
+    /**
+     * An alternative {@link ReuseStrategy} that allows swapping the stored analyzer components when they change.
+     * This is used to change e.g. token filters in search time analyzers.
+     */
+    private static final ReuseStrategy UPDATE_STRATEGY = new ReuseStrategy() {
+        @Override
+        public TokenStreamComponents getReusableComponents(Analyzer analyzer, String fieldName) {
+            ReloadableCustomAnalyzer custom = (ReloadableCustomAnalyzer) analyzer;
+            AnalyzerComponents components = custom.getStoredComponents();
+            if (components == null || custom.shouldReload(components)) {
+                custom.setStoredComponents(custom.getComponents());
```

Comment:
nit: we are accessing the components twice in the above calls (once via reload). In general I like extracting variables that may be updated by other threads to make things easier to reason about, could be extract the `Analyzer#getComponents` to a variable? This would guarantee that we are seeing exactly the same instance both times. This might not fit well with the current `shouldReload` method but maybe we can remove it and just do `components != storedComponents`?